cpp/inc/bond/ext/grpc/wait_callback.h

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+
+#include <bond/core/bonded.h>
+#include <grpc++/impl/codegen/status.h>
+
+#include <boost/optional.hpp>
+
+#include <chrono>
+#include <condition_variable>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <tuple>
+
+namespace bond { namespace ext { namespace gRPC {
+
+/// @brief A callback type that can be manually waited upon.
+///
+/// The type can be used to synchronously get the result of invoking an
+/// async proxy method.
+///
+/// The wait() member function can be used to wait until the callback has
+/// been called. Then, the status() and response() member functions can be
+/// called to inspect the results.
+template <typename TResponse>
+class wait_callback
+{
+public:
+    wait_callback() : _impl(std::make_shared<impl>()) { }
+
+    using CallbackType = std::function<void(const bond::bonded<TResponse>&, const grpc::Status&)>;
+
+    /// @brief Creates a callback that will signal this wait_callback
+    /// instance.
+    ///
+    /// If multiple callbacks are created and invoked, only the results of
+    /// the first one that is invoked will be recorded.
+    ///
+    /// It is safe for callbacks to outlive the wait_callback that created
+    /// them. When they are invoked their results will be discarded.
+    CallbackType callback()
+    {
+        // Create a weak_ptr to the implementation for the callback. If all
+        // the wait_callback instances sharing the implementation go away,
+        // their callbacks can still safely be invoked: they just won't be
+        // able to lock the weak_ptr and thus won't do anything.
+        std::weak_ptr<impl> wpImpl(_impl);
+
+        return std::function<void(const bond::bonded<TResponse>&, const grpc::Status&)>(
+            [wpImpl](const bond::bonded<TResponse>& response, const grpc::Status& status)
+            {
+                std::shared_ptr<impl> impl = wpImpl.lock();
+
+                if (impl)
+                {
+                    impl->set(response, status);
+                }
+            });
+    }
+
+    /// @brief Implicitly creates a callback.
+    operator CallbackType()
+    {
+        return callback();
+    }
```

Comment:
Wouldn't be simpler and more natural to make the `wait_callback` a function object (callback) itself by adding `operator()`?